OpenRA.Mods.Cnc/FileSystem/MegFile.cs

```diff
@@ -0,0 +1,136 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using OpenRA.FileSystem;
+using OpenRA.Primitives;
+
+namespace OpenRA.Mods.Cnc.FileSystem
+{
+	/// <summary>
+	/// This class supports loading unencrypted V3 .meg files using
+	/// reference documentation from here https://modtools.petrolution.net/docs/MegFileFormat
+	/// </summary>
+	public class MegV3Loader : IPackageLoader
+	{
+		const int HeaderFileNameTableStartOffset = 0x18;
+		const uint UnencryptedMegFlags = 0xFFFFFFFF;
+		const uint MegID = 0x3F7D70A4;
+
+		public bool TryParsePackage(Stream s, string filename, OpenRA.FileSystem.FileSystem context, out IReadOnlyPackage package)
+		{
+			var position = s.Position;
+
+			var id1 = s.ReadUInt32();
+			var id2 = s.ReadUInt32();
+
+			s.Position = position;
+
+			if (id1 != UnencryptedMegFlags || id2 != MegID)
+			{
+				package = null;
+				return false;
+			}
+
+			package = new MegFile(s, filename);
+			return true;
+		}
+
+		public sealed class MegFile : IReadOnlyPackage
+		{
+			readonly Stream s;
+
+			readonly List<string> filenames = new List<string>();
+
+			readonly Dictionary<string, (uint Offset, int Length)> contents = new Dictionary<string, (uint Offset, int Length)>();
+
+			public MegFile(Stream s, string filename)
+			{
+				Name = filename;
+				this.s = s;
+
+				var id1 = s.ReadUInt32();
+				var id2 = s.ReadUInt32();
+
+				if (id1 != UnencryptedMegFlags || id2 != MegID)
+					throw new Exception("Invalid file signature for meg file");
+
+				var dataStart = s.ReadUInt32();
+				var numFilenames = s.ReadUInt32();
+				var numFiles = s.ReadUInt32();
+				var fileNameTableSize = s.ReadUInt32();
+
+				// The file names are an indexed array of strings
+				for (var i = 0; i < numFilenames; i++)
+				{
+					var length = s.ReadUInt16();
+					filenames.Add(s.ReadASCII(length));
+				}
+
+				// The header indicates where we should be, so verify it
+				if (s.Position != fileNameTableSize + HeaderFileNameTableStartOffset)
```

Comment:
I'm not particularly a fan of hardcoding an unnecessary magic constant for `HeaderFileNameTableStartOffset`. Can we please use the stream position before the header like I suggested above?